### ISSUE TITLE
__FILE__ returns lowercase even though the files have title and camel case names. But only in the encrypted version of the extension.

### ISSUE THREAD

**Issue by FrancisGerard** [SketchUp/api-issue-tracker#130]
SketchUp Pro 2018
OS Platform: Windows 10

My extension WinDoor+ came back as approved from validation but when I tried out the downloaded encrypted code, none of the menus loaded. I checked the value of a module variable I used to collect the names of the extension files as they loaded and found that the encrypted code converted title case names to lower case only. My unecrypted code does not do this. Is this an encryption bug or should I defensively assumed that case sensitivity is not always a given?

This is the code I use to fill an array with the names of files as they load. It is at the end of each file just before the extension module end .
```ruby
this_file = __FILE__
@files_loaded << File.basename(this_file, File.extname(this_file))
```
I added a getter method to the module to see what was in @files_loaded and found that all the file names were in lower case in the downloaded version of the extension but not in my unencrypted version.

**Comment by jeroentheuns** [comment]
Hi,
Ruby __FILE__ has a UTF-8 encoding issues. Thomas told us about this during devcamp last week. Maybe ask him as i’m not fully aware of the exact details.

Jeroen

**Comment by Eneroth3** [comment]
You should always to something similar to this when using `__FILE__` to account for non-basic-lating characters in the username.

    path = __FILE__
    path.force_encoding("UTF-8") if path.respond_to?(:force_encoding)

However that isn't (as far as I can see) related to the constant being all lowercase.

**Comment by Eneroth3** [comment]
I've done some testing and it appears `__FILE__ `  isn't defined by the file name on disk but the name the file was loaded by.

Test file:
```ruby
# D:\Sketchup Plugins\Working Dir\file case\test.rb
p __FILE__
```

Run from Ruby Console (with SketchUp restart in between calls to require)
```ruby

load 'D:\Sketchup Plugins\Working Dir\file case\test.rb'
# Prints: "D:\Sketchup Plugins\Working Dir\file case\test.rb"

load 'D:\Sketchup Plugins\Working Dir\file case\TEST.rb'
# Prints: "D:\Sketchup Plugins\Working Dir\file case\TEST.rb"

Sketchup.require 'D:\Sketchup Plugins\Working Dir\file case\test.rb'
# Prints: "D:\Sketchup Plugins\Working Dir\file case\test.rb"

Sketchup.require 'D:\Sketchup Plugins\Working Dir\file case\TEST.rb'
# Prints: "D:\Sketchup Plugins\Working Dir\file case\TEST.rb"

```

Maybe Sketchup.require makes the path lowercase when loading encrypted files?

**Comment by FrancisGerard** [comment]
Thank you for doing this and for the advice to force UTF-8. It is interesting.

I have decided not to set the array values in my module variable to actual file names as they load. But rather to place a string in the array to represent the file having loaded. This gets away from what the file is called. Sketchup changes the file extension to .rbe when it is encrypted anyway so the name cannot be used without stripping that away.

**Comment by Eneroth3** [comment]
What are you using the file names for anyway? Is it a load guard for avoding duplicated menus after plugin reload?

This is what I do for load guard. No file names or other extension specific variable names needed.

```ruby
Module MyPlugin

  @loaded = false
  unless @loaded
    @loaded = true
    # Do stuff...
  end

end
```

**Comment by FrancisGerard** [comment]
Yes, i have 3 different files, each with their own menus,  so I need more than just true or false. My @loaded is an array that i place a different string into for each successfully loaded file.

**Comment by FrancisGerard** [comment]
It works the way it is but I'll think about your suggestion. It is certainly easier to add functionality to the menu if it is in the same file as the extension code.

**Comment by taustin73** [comment]
Logged SU-41960

**Comment by elsiehupp** [comment]
I'm pretty sure this is actually a bug with the `SketchUp.require(path)` method, in that the method is broken on case-sensitive APFS and causes _first-party_ SketchUp extensions to break. ~~[I found a workaround, though!](https://forums.sketchup.com/t/343138)~~

~~**EDIT:** Actually, no, the workaround didn't work. SketchUp just doesn't see the symlinked Plugins directory, so none of the plugins fail to load... because SketchUp doesn't even try to load them.~~

~~**EDIT 2:** Actually, yes, it does work, but only if you create a symlink in the terminal using, e.g.:~~

```sh
mv ~/Library/Application Support/SketchUp 2025/SketchUp/Plugins /Volumes/SketchUp/Plugins
ln -s /Volumes/SketchUp/Plugins ~/Library/Application Support/SketchUp 2025/SketchUp/Plugins
```

~~My mistake was being lazy and creating an "alias" in the Finder, which apparently isn't the same thing as a symlink.~~

~~**EDIT 3:** With the symlink to the Plugins folder, SketchUp doesn't try to load extensions, but the Extension Manager still sees them and thinks they've been loaded. So maybe using a proper symlink with the entire `~/Library/Application Support/SketchUp 2025` folder might work. I'm tired, so I'll try this later.~~

**Edit 4:** I think I figured out a workaround that actually works.

First, yes, I created an APFS volume at `/Volumes/SketchUp` with a case-insensitive filesystem. On this volume, I created a folder at `/Volumes/SketchUp/Applications` and a nested pair of folders at `/Volumes/SketchUp/Library/Application Support` (because of potential name collisions).

Then, after running SketchUp 2025 at least once (so it has an Application Support folder), I moved `~/Library/Application Support/SketchUp 2025` to `/Volumes/SketchUp/Library/Application Support/SketchUp 2025` ***and*** I moved `/Applications/SketchUp 2025` to `/Volumes/SketchUp/Applications/SketchUp 2025`, then created a symlink as follows:

```sh
ln -s /Volumes/SketchUp/Library/Application\ Support/SketchUp\ 2025 ~/Library/Application\ Support/SketchUp\ 2025
```

The reason I thought to move the application itself onto the case-insensitive volume was that when I did `puts $"` a number of the paths were within `SketchUp.app`, and `SketchUp.app` had been on my boot volume, which was case-sensitive.

For a visual representation, the contents of the volume now look like so:

<img width="449" height="421" alt="Image" src="https://github.com/user-attachments/assets/0616306f-e673-4f1a-905c-6485e8900e28" />

> _**Note:** The reason I used this folder structure was that just putting both the application and the application-support folder in the volume’s root cause a name collision, and the Finder wanted to merge the two folders, which kind of seemed like a bad idea._
> 
> _The precise names I used for the folders aren’t actually meaningful to the application itself; they are just there to help remind me, the end user, which is which._

Putting ***both*** the Application Support folder ***and*** the application itself on the case-insensitive volume seems to have done the trick.

One weird thing perhaps worth mentioning, though, is that the application didn't seem to recognize the extensions until I installed or uninstalled an extension, which I guess forced the application to try and resolve the extensions again. The way this manifested in the interface was that before I did so the extensions in the Extension Manager all had generic icons and lacked their full names, whereas after I installed or uninstalled an extension the extensions in the Extension manager displayed with the correct icons and display names.

**Comment by DanRathbun** [comment]
> I'm pretty sure this is actually a bug with the `SketchUp.require(path)` method, in that the method is broken on case-sensitive APFS ...

AFAIK, from memory, SketchUp employees have said (in the past) that case-sensitive file systems are unsupported with respect to extensions.

I did a quick test and it seems that recently a warning is now output to the console thus:
```ruby
# SketchUp's "out of the box" included Sandbox extension is already loaded ...
Sketchup.require("su_sandbox.rb")
=> false
Sketchup.require("SU_sandbox.rb")
#=> Extension loader file C:/Users/Dan/AppData/Roaming/SketchUp/SketchUp 2026/SketchUp/Plugins/SU_sandbox.rb
#=> can not be loaded because an extension loader file with the same name was previously loaded.
=> false
```
This indicates that [`Sketchup::require`](https://ruby.sketchup.com/Sketchup.html#require-class_method) is **hard-coded** for case-insensitive filesystems.

----

@thomthom @taustin73 @CraigTrickett @ChrisFullmer Is there any **official** policy posted anywhere that states support (or not) with regard to case-sensitivity of the user's file system ?

**ADD**:

> In 2017, **Barry Janzen**, Quality Assurance (QA) Engineer at SketchUp/Trimble, said ...
>_"We don't support case-sensitive filesystems on ANY version of SketchUp."_
 — [Forum Link: Mac OS X 10.11.6, SketchUp Make 2017 Crashes Immediately - (post #32)](https://forums.sketchup.com/t/mac-os-x-10-11-6-sketchup-make-2017-crashes-immediately-when-start-using-sketchup-is-clicked/35098/32)
----

Users describe symlink workatounds in 
https://forums.sketchup.com/t/sketchup-on-mac-with-case-sensitive-filesystem/136770

**Comment by elsiehupp** [comment]
tl;dr the very simple workaround is to move **both** `SketchUp.app` and its Application Support folder to a case-insensitive APFS volume:

1. Run SketchUp at least once, so that it creates an Application Support folder.
2. Create a case-insensitive APFS volume called, e.g., `/Volumes/SketchUp`. See [the Apple KB article on APFS volumes](https://support.apple.com/guide/disk-utility/add-delete-or-erase-apfs-volumes-dskua9e6a110/mac). (There's probably a way to do this using the Terminal if you want to be clever about it.)
3. Open Terminal and run the following commands (see screenshot for what this should look like):
    <img width="449" height="421" alt="Image" src="https://github.com/user-attachments/assets/0616306f-e673-4f1a-905c-6485e8900e28" />
```sh
mkdir /Volumes/SketchUp/Applications
mkdir -p /Volumes/SketchUp/Library/Application\ Support
mv /Applications/SketchUp\ 2025 /Volumes/SketchUp/Applications/SketchUp\ 2025
mv ~/Library/Application\ Support/SketchUp\ 2025 Volumes/SketchUp/Library/Application\ Support/SketchUp\ 2025
ln -s /Volumes/SketchUp/Library/Application\ Support/SketchUp\ 2025 /Users/elsiehupp/Library/Application\ Support/SketchUp\ 2025
```
4. Launch SketchUp and open Extensions > Extension Manager. If the extensions have generic icons:
    a. Open Extensions > Extension Warehouse and install any extension at random.
    b. Go back to the Extension Manager and uninstall the extension you just installed if you don't actually need it.
    This install/uninstall will force SketchUp to reload all extensions.

It seems that you may need to do the install/uninstall rigmarole every single time you launch SketchUp, which is annoying, but it's better than having extensions simply not work.

Also, note, obviously `SketchUp 2025` in the Terminal commands is for a particular version; you will need to change each occurrence to, e.g., `SketchUp 2026` if that's the version you're working with.
